File: src/ESedexCEPChecking.java
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class ESedexCEPChecking {
	private static final int MINIMUM_LENGTH = 9 + 1 + 9 + 1;
	private final List<ESedexCEPChecking.CEPRange> cepRanges = new ArrayList<ESedexCEPChecking.CEPRange>();

	public ESedexCEPChecking() throws Exception {
		this.loadAllCEPRanges();
	}

	private void loadAllCEPRanges() throws Exception {
		final File ceps = new File(new File("."), "ceps.txt");
		final Scanner scanner = new Scanner(ceps);
		try {
			while (scanner.hasNextLine()) {
				String line = scanner.nextLine().trim();

				if (line.startsWith("#") || line.isEmpty() || line.length() < ESedexCEPChecking.MINIMUM_LENGTH) {
					continue;
				}

				final int lastCEP = Integer.parseInt(line.substring(line.lastIndexOf(" ") + 1, line.length()).replace("-", ""));
				line = line.substring(0, line.lastIndexOf(" "));

				final int priorLastCEP = Integer.parseInt(line.substring(line.lastIndexOf(" ") + 1, line.length()).replace("-", ""));

				final boolean onlyPostage = line.contains("*");

				this.cepRanges.add(new CEPRange(priorLastCEP, lastCEP, onlyPostage));
			}
		} catch (final Exception e) {
			throw e;
		} finally {
			scanner.close();
		}
	}

	public CEPStatus check(String stringCEP) {
		if (stringCEP == null) {
			throw new NullPointerException("stringCEP cannot be null");
		}
		stringCEP = stringCEP.trim().replace("-", "");
		if (stringCEP.length() != 8) {
			throw new IllegalArgumentException("invalid CEP '" + stringCEP + "'");
		}
		final int cep = Integer.parseInt(stringCEP);

		for (final CEPRange cepRange : this.cepRanges) {
			if (cepRange.isAvailable(cep)) {
				return cepRange.isOnlyPostage() ? CEPStatus.YES_ONLY_POSTAGE : CEPStatus.YES;
			}
		}
		return CEPStatus.NO;
	}

	private static class CEPRange {
		private final long min;
		private final long max;
		private final boolean onlyPostage;

		public CEPRange(final long min, final long max, final boolean onlyPostage) {
			this.min = min;
			this.max = max;
			this.onlyPostage = onlyPostage;
		}

		public boolean isAvailable(final int cep) {
			return cep >= this.min && cep <= this.max;
		}

		public boolean isOnlyPostage() {
			return this.onlyPostage;
		}
	}
}

File: esedex-cep-checking.rb

class CEPRange
  def initialize min, max, hasOnlyPostage
    @range, @hasOnlyPostage = Range.new( formatCEP(min), formatCEP(max) ), hasOnlyPostage
  end

  def include? cep
    @range.include? formatCEP(cep)
  end

  def onlyPostage?
    @hasOnlyPostage
  end
  
  private
  def formatCEP cep
    cep.gsub('-', '').to_i
  end
end

class CEPChecking
  def initialize ceps_files='./ceps.txt'
    @ranges = []
    file = File.open(ceps_files);
    begin
      file.readlines.each do |line|
        line = line.strip.chomp
        addRange line unless line.size < 20 or line.start_with? '#'
      end
    ensure
      file.close
    end
  end

  def check cep
    @ranges.each do |r|
      if r.include? cep
        return r.onlyPostage? ? :yes_only_postage : :yes
      end
    end
    :no
  end

  private
  def addRange line
    lastCEP = line[line.rindex(' '), line.size].strip

    priorLastCEP = line[0, line.rindex(' ')]
    priorLastCEP = priorLastCEP[priorLastCEP.rindex(' '), priorLastCEP.size]

    @ranges << CEPRange.new( priorLastCEP, lastCEP, line.include?('*') )
  end
end

puts "should be yes_only_postage => #{CEPChecking.new.check('19400000')}"
puts "should be yes_only_postage => #{CEPChecking.new.check('19409999')}"
puts "should be yes => #{CEPChecking.new.check('01000-000')}"
puts "should be yes => #{CEPChecking.new.check('09999999')}"
puts "should be no => #{CEPChecking.new.check('65884-884')}"
